LiS/man/man4/echo.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: echo.4.man,v 0.9.2.3 2008-04-29 08:33:18 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:18 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: echo.4.man,v $
.\" Revision 0.9.2.3  2008-04-29 08:33:18  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH ECHO 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B echo
\- echo \fISTREAMS\fP device
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <unistd.h>
.br
.B #include <stropts.h>
.HP 12
.BI "int " fd " = open(" /dev/echo ", " oflag );
.\"
.\"
.SH DESCRIPTION
.PP
.I fd
is a file descriptor for an echo stream opened using the
.BR open (2)
system call on one of the following devices:
.TP
.RB \*(lq /dev/echo \*(rq
This device is accessible to all users.
.PP
The
.B echo
device is characterized by the following behaviour:
.IP \(em \w'\(em\(em'u
Most 
.I STREAMS
messages written to the driver's write queue are echoed back to the stream
head read queue.  These messages include:
.BR M_DATA,
.BR M_PROTO,
and
.BR M_PCPROTO .
.\"
.\"
.SH IOCTLS
.PP
.B echo
does not support any I/O controls of its own and all I/O controls delivered
will return an
.IR errno (3)
of
.BR [ EOPNOTSUPP ].
.\"
.\"
.SH USAGE
.PP
The
.B echo
device can be opened when no other suitable device exists and modules may be
pushed on the
.B echo
device.
.PP
The primary use of the
.B echo
device is the
.I STREAMS
verification utility
.BR strvf (8).
.\"
.\"
.SH NOTICES
.PP
The
.B echo
device is used on
.IR HP-UX \(rg
systems in support of the
.BR strvf (8)
utility.
.B Linux Fast-STREAMS
provides this device in support of
.BR strvf (8)
and for compatibility with
.IR HP-UX \(rg
and
.IR OSF/1 \(rg.
.\"
.\"
.SH IMPLEMENTATION
.PP
.B echo
is implemented as a regular
.I STREAMS
pseudo-device driver.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH DEVICES
.PP
.B /dev/echo
.\"
.\"
.SH "SEE ALSO"
.PP
.BR open (2),
.IR errno (3),
.BR nuls (4),
.BR spx (4),
.BR strvf (8).
.\"
.\"
.SH BUGS
.PP
.B echo
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.IR HP-UX \(rg.
This is an
.IR HP-UX \(rg-
and
.IR OSF/1 \(rg-specific
device.
.[
hpuxspg
.]
.[
osfspg
.]
.B echo
is provided in
.B Linux Fast-STREAMS
for compatibility with
.IR HP-UX \(rg
and
.IR OSF/1 \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IR HP-UX \(rg
and
.IR OSF/1 \(rg
do not document all behaviour.
.IP \(em
Use for purposes other than
.BR strvf (8)
are not necessarily compatible.
.IP \(em
For an
.IR AIX \(rg
and
.IR UnixWare \(rg
equivalent, see
.BR spx (4).
.\"
.\"
.SH CONFORMANCE
.PP
.IR HP-UX \(rg
and
.IR OSF/1 \(rg
documentation.
.[
hpuxdoc
.]
.[
osfdoc
.]
.\"
.\"
.SH HISTORY
.PP
.B echo
first appeared in
.IR HP-UX \(rg
and
.IR OSF/1 \(rg?
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/echo.c
/*****************************************************************************

 @(#) $RCSfile: echo.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: echo.c,v $
 Revision 0.9.2.15  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 0.9.2.14  2007/08/15 04:57:58  brian
 - GPLv3 updates

 Revision 0.9.2.13  2007/08/14 10:46:56  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: echo.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] =
    "$RCSfile: echo.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $";

#include <linux/autoconf.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#ifdef LFS
#include <sys/strsubr.h>
#endif
#include <sys/ddi.h>

#ifdef LFS
#include "sys/config.h"
#endif

#define ECHO_DESCRIP	"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define ECHO_COPYRIGHT	"Copyright (c) 1997-2008 OpenSS7 Corporation.  All Rights Reserved."
#define ECHO_REVISION	"LfS $RCSfile: echo.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $"
#define ECHO_DEVICE	"SVR 4.2 STREAMS Echo (ECHO) Device"
#define ECHO_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define ECHO_LICENSE	"GPL"
#define ECHO_BANNER	ECHO_DESCRIP	"\n" \
			ECHO_COPYRIGHT	"\n" \
			ECHO_REVISION	"\n" \
			ECHO_DEVICE	"\n" \
			ECHO_CONTACT	"\n"
#define ECHO_SPLASH	ECHO_DEVICE	" - " \
			ECHO_REVISION	"\n"

#if defined LIS && defined MODULE
#define CONFIG_STREAMS_ECHO_MODULE MODULE
#endif

#ifdef CONFIG_STREAMS_ECHO_MODULE
MODULE_AUTHOR(ECHO_CONTACT);
MODULE_DESCRIPTION(ECHO_DESCRIP);
MODULE_SUPPORTED_DEVICE(ECHO_DEVICE);
MODULE_LICENSE(ECHO_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-echo");
#endif
#endif

#ifndef CONFIG_STREAMS_ECHO_NAME
#ifdef LIS
#define CONFIG_STREAMS_ECHO_NAME ECHO__DRV_NAME
#endif
#ifdef LFS
#define CONFIG_STREAMS_ECHO_NAME "echo"
#endif
#endif
#ifndef CONFIG_STREAMS_ECHO_MODID
#ifdef LIS
#define CONFIG_STREAMS_ECHO_MODID ECHO__ID
#endif
#ifdef LFS
#error CONFIG_STREAMS_ECHO_MODID must be defined.
#endif
#endif
#ifndef CONFIG_STREAMS_ECHO_MAJOR
#ifdef LIS
#define CONFIG_STREAMS_ECHO_MAJOR ECHO__CMAJOR_0
#endif
#ifdef LFS
#error CONFIG_STREAMS_ECHO_MAJOR must be defined.
#endif
#endif

modID_t modid = CONFIG_STREAMS_ECHO_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module id number for ECHO driver. (0 for auto allocation)");

#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_ECHO_MODID));
MODULE_ALIAS("streams-driver-echo");
#endif

major_t major = CONFIG_STREAMS_ECHO_MAJOR;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for ECHO driver. (0 for auto allocation)");

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_ECHO_MAJOR) "-*");
MODULE_ALIAS("/dev/echo");
#ifdef LFS
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_ECHO_MAJOR));
MODULE_ALIAS("/dev/streams/echo");
MODULE_ALIAS("/dev/streams/echo/*");
MODULE_ALIAS("/dev/streams/clone/echo");
#endif
#endif

#ifdef LIS
#define STRMINPSZ   0
#define STRMAXPSZ   4096
#define STRHIGH	    5120
#define STRLOW	    1024
#endif

static struct module_info echo_minfo = {
	.mi_idnum = CONFIG_STREAMS_ECHO_MODID,
	.mi_idname = CONFIG_STREAMS_ECHO_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat echo_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat echo_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

#ifdef LIS
#define _trace() while (0) { }
#define _ptrace(__x) while (0) { }
#define _printd(__x) while (0) { }
#define pswerr(__x) while (0) { }
#define _ctrace(__x) __x

#define QSVCBUSY QRUNNING

union ioctypes {
	struct iocblk iocblk;
	struct copyreq copyreq;
	struct copyresp copyresp;
};
#endif

#ifdef LIS
#define streamscall _RP
#endif

static streamscall int
echo_rput(queue_t *q, mblk_t *mp)
{
	putnext(q, mp);
	return (0);
}

static streamscall int
echo_rsrv(queue_t *q)
{
	qenable(OTHERQ(q));
	return (0);
}

static streamscall int
echo_wput(queue_t *q, mblk_t *mp)
{
	int err = 0;

	_trace();
	switch (mp->b_datap->db_type) {
	case M_FLUSH:
		_trace();
		if (mp->b_rptr[0] & FLUSHW) {
			_trace();
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
			mp->b_rptr[0] &= ~FLUSHW;
		}
		if (mp->b_rptr[0] & FLUSHR) {
			_trace();
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(RD(q), mp->b_rptr[1], FLUSHDATA);
			else
				flushq(RD(q), FLUSHDATA);
			_ctrace(qreply(q, mp));
			/* never makes it here */
			_trace();
			return (0);
		}
		_trace();
		break;
	case M_IOCTL:
	case M_IOCDATA:
	{
		union ioctypes *ioc;

		_ptrace(("received M_IOCTL or M_IOCDATA, naking it\n"));
		err = -EINVAL;

		mp->b_datap->db_type = M_IOCNAK;
		ioc = (typeof(ioc)) mp->b_rptr;
		ioc->iocblk.ioc_count = 0;
		ioc->iocblk.ioc_rval = -1;
		ioc->iocblk.ioc_error = -err;
		qreply(q, mp);
		return (0);
	}
	case M_READ:
		mp->b_wptr = mp->b_rptr;
		mp->b_datap->db_type = M_DATA;
		mp->b_flag |= MSGDELIM;
		qreply(q, mp);
		return (0);
	case M_DATA:
	case M_PROTO:
	case M_PCPROTO:
	case M_CTL:
	case M_PCCTL:
	case M_RSE:
	case M_PCRSE:
		/* Check QSVCBUSY flag in MP drivers to avoid missequencing of messages when
		   service procedure is running concurrent with put procedure. */
		if (mp->b_datap->db_type < QPCTL && (q->q_first || (q->q_flag & QSVCBUSY)
						     || !bcanputnext(OTHERQ(q), mp->b_band)))
			putq(q, mp);
		else
			qreply(q, mp);
		return (0);
	}
	freemsg(mp);
	return (0);
}

static streamscall int
echo_wsrv(queue_t *q)
{
	mblk_t *mp;
	queue_t *rq = RD(q);

	while ((mp = getq(q))) {
		if (bcanputnext(rq, mp->b_band)) {
			putnext(rq, mp);
			continue;
		}
		putbq(q, mp);
		break;
	}
	return (0);
}

typedef struct echo {
	struct echo *next;
	struct echo **prev;
	dev_t dev;
} echo_t;

static spinlock_t echo_lock = SPIN_LOCK_UNLOCKED;
static struct echo *echo_list = NULL;

/* 
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
static streamscall int
echo_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	struct echo *p, **pp = &echo_list;
	major_t cmajor = getmajor(*devp);
	minor_t cminor = getminor(*devp);

	_ptrace(("%s: opening major %hu, minor %hu, sflag %d\n", __FUNCTION__, cmajor, cminor,
		 sflag));
	if (q->q_ptr != NULL) {
		_printd(("%s: stream is already open\n", __FUNCTION__));
		return (0);	/* already open */
	}
	if (sflag == MODOPEN || WR(q)->q_next) {
		_printd(("%s: cannot open as module\n", __FUNCTION__));
		return (ENXIO);	/* can't open as module */
	}
	if (!(p = kmem_alloc(sizeof(*p), KM_NOSLEEP))) {	/* we could sleep */
		_printd(("%s: could not allocate private structure\n", __FUNCTION__));
		return (ENOMEM);	/* no memory */
	}
	bzero(p, sizeof(*p));
	switch (sflag) {
	case CLONEOPEN:
		_printd(("%s: clone open\n", __FUNCTION__));
		if (cminor < 1)
			cminor = 1;
	case DRVOPEN:
	{
		major_t dmajor = cmajor;

		_printd(("%s: driver open\n", __FUNCTION__));
		if (cminor < 1) {
			_printd(("%s: attempt to open minor zero non-clone\n", __FUNCTION__));
			return (ENXIO);
		}
		spin_lock(&echo_lock);
		for (; *pp && (dmajor = getmajor((*pp)->dev)) < cmajor; pp = &(*pp)->next) ;
		for (; *pp && dmajor == getmajor((*pp)->dev) &&
		     getminor(makedevice(cmajor, cminor)) != 0; pp = &(*pp)->next) {
			minor_t dminor = getminor((*pp)->dev);

			if (cminor < dminor)
				break;
			if (cminor == dminor) {
				if (sflag == CLONEOPEN)
					cminor++;
				else {
					spin_unlock(&echo_lock);
					kmem_free(p, sizeof(*p));
					pswerr(("%s: stream already open!\n", __FUNCTION__));
					return (EIO);	/* bad error */
				}
			}
		}
		if (getminor(makedevice(cmajor, cminor)) == 0) {	/* no minors left */
			spin_unlock(&echo_lock);
			kmem_free(p, sizeof(*p));
			_printd(("%s: no minor devices left\n", __FUNCTION__));
			return (EBUSY);	/* no minors left */
		}
		p->dev = *devp = makedevice(cmajor, cminor);
		if ((p->next = *pp))
			p->next->prev = &p->next;
		p->prev = pp;
		*pp = p;
		q->q_ptr = OTHERQ(q)->q_ptr = p;
		spin_unlock(&echo_lock);
		qprocson(q);
		_printd(("%s: opened major %hu, minor %hu\n", __FUNCTION__, cmajor, cminor));
		return (0);
	}
	}
	pswerr(("%s: bad sflag %d\n", __FUNCTION__, sflag));
	return (ENXIO);
}

static streamscall int
echo_close(queue_t *q, int oflag, cred_t *crp)
{
	struct echo *p;

	_trace();
	if ((p = q->q_ptr) == NULL) {
		pswerr(("%s: already closed\n", __FUNCTION__));
		return (0);	/* already closed */
	}
	qprocsoff(q);
	spin_lock(&echo_lock);
	if ((*(p->prev) = p->next))
		p->next->prev = p->prev;
	p->next = NULL;
	p->prev = &p->next;
	q->q_ptr = OTHERQ(q)->q_ptr = NULL;
	spin_unlock(&echo_lock);
	_printd(("%s: closed stream with read queue %p\n", __FUNCTION__, q));
	return (0);
}

static struct qinit echo_rqinit = {
	.qi_putp = echo_rput,
	.qi_srvp = echo_rsrv,
	.qi_qopen = echo_open,
	.qi_qclose = echo_close,
	.qi_minfo = &echo_minfo,
	.qi_mstat = &echo_rstat,
};

static struct qinit echo_wqinit = {
	.qi_putp = echo_wput,
	.qi_srvp = echo_wsrv,
	.qi_minfo = &echo_minfo,
	.qi_mstat = &echo_wstat,
};

static struct streamtab echo_info = {
	.st_rdinit = &echo_rqinit,
	.st_wrinit = &echo_wqinit,
};

#ifdef LFS
static struct cdevsw echo_cdev = {
	.d_name = CONFIG_STREAMS_ECHO_NAME,
	.d_str = &echo_info,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};
#endif

#ifdef CONFIG_STREAMS_ECHO_MODULE
static
#endif
int __init
echo_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_ECHO_MODULE
	printk(KERN_INFO ECHO_BANNER);
#else
	printk(KERN_INFO ECHO_SPLASH);
#endif
	echo_minfo.mi_idnum = modid;
#ifdef LFS
	if ((err = register_strdev(&echo_cdev, major)) < 0)
		return (err);
#endif
#ifdef LIS
	if ((err = lis_register_strdev(major, &echo_info, 255, CONFIG_STREAMS_ECHO_NAME)) < 0)
		return (err);
#endif
	if (major == 0 && err > 0)
		major = err;
#ifdef LIS
	if ((err = lis_register_driver_qlock_option(major, LIS_QLOCK_NONE)) < 0) {
		lis_unregister_strdev(major);
		return (err);
	}
#endif
	return (0);
};

#ifdef CONFIG_STREAMS_ECHO_MODULE
static
#endif
void __exit
echo_exit(void)
{
#ifdef LFS
	unregister_strdev(&echo_cdev, major);
#endif
#ifdef LIS
	lis_unregister_strdev(major);
#endif
};

#ifdef CONFIG_STREAMS_ECHO_MODULE
module_init(echo_init);
module_exit(echo_exit);
#endif
